presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp

```diff
@@ -986,9 +986,9 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
       VELOX_UNSUPPORTED("Unsupported aggregation step");
   }
 
-  bool streamable = !node->preGroupedVariables.empty() &&
-      node->groupingSets.groupingSetCount == 1 &&
-      node->groupingSets.globalGroupingSets.empty();
+  // TODO karteekmurthys@ Re-enable after
+  // fixing:https://github.com/prestodb/presto/issues/22585
+  bool streamable = false;
```

Comment:
Is this safe? If planner planned streaming agg it means it has some assumptions that are wrong. Failing the query might be safer.

Reply:
@mbasmanova : streamable just sets preGroupedVariables for Velox AggregationNode.

https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp#L1010 

Presto planner doesn't really specify streaming agg. The streaming agg choice is made in Velox planner based on preGrouped variables https://github.com/facebookincubator/velox/blob/main/velox/exec/LocalPlanner.cpp#L499

So I think blanking out preGrouped variables is sufficient for letting the aggregation proceed with HashAggregation.

Let me know if you think otherwise.

Reply:
@aditi-pandit Aditi, you are correct that replacing streaming agg with hash agg is safe (just less efficient). However, my concern is that there could be other changes to the plan made by the optimizer based on wrong assumptions that led it to specify pre-grouped fields "incorrectly". In general, it is not safe to "question" the optimizer. However, here we say that we've seen cases where agg with pre-grouped keys was a result of optimizer making incorrect assumptions, hence, we are not following that choice. What I'm saying is that if part of the plan cannot be trusted, I don't think we can trust the rest of the plan. Hence, I'm saying that it would be safer to reject the whole plan and fail.

Reply:
@mbasmanova : The pre-Grouping seemed broken only with nestedLoopJoin, but okay otherwise. 

The current fix to disable the streaming agg always is also aggressive to avoid fine-tuning plan patterns for aggregations over Nested Loop joins.

I feel rejecting the whole plan would be super conservative. But if you want that behavior in Meta production I won't question it. We will likely change the behavior internally at IBM.

Reply:
My understanding is that a proper fix will be ready in Velox soon, so hopefully no Presto mitigation will be needed at all. But FWIW, I don't think there are other things relying on the pregrouped aggregation variables, so I think that aspect of just converting to a regular aggregation that ignores the pregrouped variables is pretty safe for a short term fix (wouldn't rely on it longer term because of course things can change). 

The riskier part that neither PR addresses is whether other optimizations unrelated to aggregation are also depending on nested loop join maintaining the ordering of the input. Would we plan a merge join after a nested loop join because we think the data is ordered?  Or drop an order by after a nested loop join because we think the data is already ordered?  If we'd like to be conservative/thorough, the safest thing to do is rather than specifically handling aggregation, fail all queries that are using NestedLoopJoin, which is the source of the incorrect optimization.

Reply:
@karteekmurthys Just discussed with @aditi-pandit. We agreed to fail the queries in this case, similar to what @rschlussel and @mbasmanova suggested. Lets change it to throw an exception instead of disabling streaming aggregation.

Reply:
We currently use NestedLoopJoin when there is no Join criteria sepcified. If we totally disable it, we don't have alternative way to implement this situation.

```
if (node->criteria.empty()) {
    const bool isNestedJoinType = core::isInnerJoin(joinType) ||
        core::isLeftJoin(joinType) || core::isRightJoin(joinType) ||
        core::isFullJoin(joinType);
    if (isNestedJoinType) {
      return std::make_shared<core::NestedLoopJoinNode>(
          node->id,
          joinType,
          node->filter ? exprConverter_.toVeloxExpr(*node->filter) : nullptr,
          toVeloxQueryPlan(node->left, tableWriteInfo, taskId),
          toVeloxQueryPlan(node->right, tableWriteInfo, taskId),
          toRowType(node->outputVariables, typeParser_));
    }
    VELOX_UNSUPPORTED(
        "JoinNode has empty criteria that cannot be "
        "satisfied by NestedJoinNode or HashJoinNode");
  }
```

@mbasmanova Can you confirm what is the consequence of disabling Nestedloopjoin for Meta's workloads?

Reply:
If we fail for this case it means that a small percentage of queries will start to fail in our native workers and need to run in Java clusters (and even though it is a small percentage, you are right that it's still a much larger impact than just failing for streaming aggregations, and it's not impact we want to take on lightly).  still we think it's the right think to do for the short term because even though most nested loop join queries are probably fine, we can't guarantee the correctness of queries that have nested loop joins in them.

Reply:
@rschlussel : NLJ is used for any join where there isn't any equality condition between join columns. Sounds like you have have very little traffic for this.

@karteekmurthys : We should do the conservative approach in the open source project. We will have to internally enable this for us at IBM separately. Please follow up with Ethan for that.